File: src/main/java/ufcg/entities/Grupo.java
package ufcg.entities;

import java.util.ArrayList;
import java.util.List;

public class Grupo {
    private String nome;
    private List<Aluno> membros;
    private int tamanho = 0;

    public Grupo(String nome, int tamanho) {
        if(tamanho <= 0) throw new IllegalArgumentException("Tamanho de grupo precisa ser maior que 0");
        this.nome = nome;
        this.membros = new ArrayList<Aluno>();
        this.tamanho = tamanho;
    }

    public Grupo(String nome) {
        this.nome = nome;
        this.membros = new ArrayList<Aluno>(null);
    }

    public void adicionarAluno(Aluno aluno) throws Exception {
        if(tamanho == 0 || membros.size() < tamanho) {
            membros.add(aluno);
        } else {
            throw new Exception("GRUPO CHEIO");
        }
    }

    public boolean isAlunoEmGrupo(Aluno aluno) {
        for(Aluno al : membros) if(aluno.getMatricula().equals(al.getMatricula())) return true;
        return false;
    }

    public String getNome() {
        return this.nome;
    }

    public List<Aluno> getMembros() {
        return this.membros;
    }
}

File: rust/src/models/grupo.rs
use crate::models::aluno::Aluno;

pub struct Grupo<> {
    nome: String,
    tamanho: usize,
    alunos: Vec<Aluno>
}

impl Grupo {
    pub fn new(inome: String, itamanho: usize) -> Self {
        Grupo {
            nome: inome,
            tamanho: itamanho,
            alunos: Vec::new()
        }
    }

    pub fn get_nome(&self) -> &String{
        &self.nome
    }

    pub fn is_aluno_alocado(&self, imatricula: &String) -> bool {
        for aluno in &self.alunos {
            if aluno.get_matricula() == imatricula {
                return true;
            }
        }
        false
    }

    pub fn alocar_aluno(&mut self, aluno: Aluno) {
        if self.tamanho == 0 || self.alunos.len() < self.tamanho {
            if !self.is_aluno_alocado(aluno.get_matricula()) {
                self.alunos.push(aluno);
            }
        } else {
            println!("Grupo Cheio!");
        }

    }
}